frontend/public/components/search.tsx

```diff
@@ -0,0 +1,257 @@
+    return <LoadingBox />;
+  }
+
+  const componentLoader = resourceListPages.get(referenceForModel(kindObj), () =>
+    Promise.resolve(DefaultPage),
+  );
+  const ns = kindObj.namespaced ? namespace : undefined;
+
+  return (
+    <AsyncComponent
+      loader={componentLoader}
+      namespace={ns}
+      selector={selector}
+      nameFilter={nameFilter}
+      kind={kindObj.crd ? referenceForModel(kindObj) : kindObj.kind}
+      showTitle={false}
+      hideTextFilter
+      autoFocus={false}
+      mock={mock}
+    />
+  );
+});
+
+const SearchPage_: React.FC<SearchProps> = (props) => {
+  const [selectedItems, setSelectedItems] = React.useState([]);
+  const [collaspedState, setCollaspedState] = React.useState(new Set([]));
+  const [selector, setSelector] = React.useState(selectorFromString(''));
+  const [nameFilter, setNameFilter] = React.useState([]);
+  const [labelFilter, setLabelFilter] = React.useState([]);
+
+  const { namespace, noProjectsAvailable } = props;
+
+  // Set state variables from the URL
+  React.useEffect(() => {
+    let kind: string, q: string, name: string;
+
+    if (window.location.search) {
+      const sp = new URLSearchParams(window.location.search);
+      kind = sp.get('kind');
+      q = sp.get('q');
+      name = sp.get('name');
+    }
+    // Ensure that the "kind" route parameter is a valid resource kind ID
+    kind = kind || '';
+    if (kind !== '') {
+      setSelectedItems(kind.split(','));
+    }
+    const tags = split(q || '');
+    const nameTags = split(name || '');
+    const validTags = _.reject(tags, (tag) => requirementFromString(tag) === undefined);
+    const validNameTags = _.reject(nameTags, (tag) => requirementFromString(tag) === undefined);
+    setSelector(selectorFromString(validTags.join(',')));
+    setLabelFilter(validTags);
+    setNameFilter(validNameTags);
+  }, []);
+
+  const updateSelectedItems = (selection: string) => {
+    const updateItems = new Set([...selectedItems]);
+    updateItems.has(selection) ? updateItems.delete(selection) : updateItems.add(selection);
+    setSelectedItems([...updateItems]);
+    setQueryArgument('kind', [...updateItems].join(','));
+  };
+
+  const clearSelectedItems = () => {
+    setSelectedItems([]);
+    setQueryArgument('kind', '');
+  };
+
+  const clearNameFilter = () => {
+    setNameFilter([]);
+    setQueryArgument('name', '');
+  };
+
+  const clearLabelFilter = () => {
+    setLabelFilter([]);
+    setSelector(selectorFromString(''));
+    setQueryArgument('q', '');
+  };
+
+  const clearAll = () => {
+    setSelectedItems([]);
+    setLabelFilter([]);
+    setNameFilter([]);
+    setSelector(selectorFromString(''));
+    setQueryArgument('kind', '');
+    setQueryArgument('q', '');
+    setQueryArgument('name', '');
+  };
+
+  const isKindExpanded = (kindView: string) => {
+    return collaspedState.has(kindView);
+  };
+
+  const toggleKindExpanded = (kindView: string) => {
+    const newCollasped = new Set(collaspedState);
```

Comment:
We can't we just store the `Set` in state?

Reply:
I am just making a copy there so I can then modify the state.  The collaspedState is now a Set.

Reply:
> I am just making a copy there so I can then modify the state. The collaspedState is now a Set.

OK, I see. We might make this an `ImmutableSet`, but this is fine.